src/icrar/leap-accelerate/cuda/MetaDataCuda.cu

```diff
@@ -0,0 +1,192 @@
+        && dlm_ra == rhs.dlm_ra
+        && dlm_dec == rhs.dlm_dec;
+    }
+
+    MetaDataCudaHost::MetaDataCudaHost(const MetaData& metadata)
+    {
+        m_constants.nantennas = metadata.nantennas;
+        m_constants.channels = metadata.channels;
+        m_constants.num_pols = metadata.num_pols;
+        m_constants.stations = metadata.stations;
+        m_constants.rows = metadata.rows;
+        m_constants.freq_start_hz = metadata.freq_start_hz;
+        m_constants.freq_inc_hz = metadata.freq_inc_hz;
+        m_constants.phase_centre_ra_rad = metadata.phase_centre_ra_rad;
+        m_constants.phase_centre_dec_rad = metadata.phase_centre_dec_rad;
+        m_constants.dlm_ra = metadata.dlm_ra;
+        m_constants.dlm_dec = metadata.dlm_dec;
+
+        init = metadata.init;
+        oldUVW = metadata.oldUVW;
+
+        avg_data = ConvertMatrix(metadata.avg_data);
+        
+        if(metadata.dd.is_initialized())
+        {
+            dd = ConvertMatrix3x3(metadata.dd.value());
+        }
+        else
+        {
+            throw std::runtime_error("dd is required");
+        }
+
+        A = ConvertMatrix(metadata.A);
+        I = ConvertMatrix<int>(metadata.I);
+        Ad = ConvertMatrix(metadata.Ad);
+
+        A1 = ConvertMatrix(metadata.A1);
+        I1 = ConvertMatrix<int>(metadata.I1);
+        Ad1 = ConvertMatrix(metadata.Ad1);
+    }
+
+    const Constants& MetaDataCudaHost::GetConstants() const
+    {
+        return m_constants;
+    }
+
+    void MetaDataCudaHost::CalcUVW(std::vector<casacore::MVuvw>& uvws)
+    {
+        this->oldUVW = uvws;
+        auto size = uvws.size();
+        uvws.clear();
+        for(int n = 0; n < size; n++)
+        {
+            auto uvw = icrar::Dot(uvws[n], dd);
+            uvws.push_back(uvw);
+        }
+    }
+
+    void MetaDataCudaHost::SetDD(const casacore::MVDirection& direction)
+    {
+        m_constants.dlm_ra = direction.get()[0] - m_constants.phase_centre_ra_rad;
+        m_constants.dlm_dec = direction.get()[1] - m_constants.phase_centre_dec_rad;
+
+        dd(0,0) = cos(m_constants.dlm_ra) * cos(m_constants.dlm_dec);
+        dd(0,1) = -sin(m_constants.dlm_ra);
+        dd(0,2) = cos(m_constants.dlm_ra) * sin(m_constants.dlm_dec);
+        
+        dd(1,0) = sin(m_constants.dlm_ra) * cos(m_constants.dlm_dec);
+        dd(1,1) = cos(m_constants.dlm_ra);
+        dd(1,2) = sin(m_constants.dlm_ra) * sin(m_constants.dlm_dec);
+
+        dd(2,0) = -sin(m_constants.dlm_dec);
+        dd(2,1) = 0;
+        dd(2,2) = cos(m_constants.dlm_dec);
+    }
+
+    void MetaDataCudaHost::SetWv()
+    {
+        double speed_of_light = 299792458.0;
+        m_constants.channel_wavelength = range(
+            m_constants.freq_start_hz,
+            m_constants.freq_inc_hz,
+            m_constants.freq_start_hz + m_constants.freq_inc_hz * m_constants.channels);
+        for(double& v : m_constants.channel_wavelength)
+        {
+            v = speed_of_light / v;
+        }
```

Comment:
That looks (I think) exactly like the implementation in `icrar::MetaData`, could that code be shared?

The code of `SetDD` is also the same I think, after removing the usage of `IPosition`-based indexing in `icrar::MetaData::SetDD`, so those could also probably be shared.

Reply:
I updated these to be methods and they probably could be private as SetWv should really be called in the constructor.

For cuda I might remove this as it may be faster to just compute number from an index rather than copy an extra dynamically sized buffer.

I move speed_of_light into a common folder later on, I'll cherry pick that change into here.

Reply:
If `SetWV` will change in the CUDA version of the code then that's fine. Regarding `SetDD`, after your changes: does it look similar enough that you could de-duplicate the code?

Reply:
Both implementations are unit tested with the same set of inputs and output to intentionally allow the optimized algorithm and class diagram to change freely, in a later PR the interface of the optimized implementation changes from the original. I should probably add documentation to some of the classes now. Basically the difference between MetaData (later becomes casalib::MetaData) and MetaDataHost (later becomes cpu::MetaData) 

casalib::MetaData is a (near) direct port of the working python code primarily for behaviour driven testing of other implementations. Once it's behaviour is the same as the python implementation there is no need to change it.

MetaDataHost/cpu::MetaData is metadata for a c++ cpu optimized implementation of leap-cal using eigen in layout mirroring the cuda implementation (free to completely change from casalib::MetaData)

Reply:
Ohh, I see, I had failed to grasp that intent the first time around, no problems then :)